llvm/include/llvm/Target/GenericOpcodes.td

```diff
@@ -1209,6 +1209,15 @@ def G_FENCE : GenericInstruction {
   let hasSideEffects = true;
 }
 
+// Generic opcode equivalent to the llvm.prefetch intrinsic.
+def G_PREFETCH : GenericInstruction {
+  let OutOperandList = (outs);
+  let InOperandList = (ins ptype0:$address, i32imm:$rw, i32imm:$locality, i32imm:$cachetype);
+  let hasSideEffects = true;
+  let mayLoad = true;
+  let mayStore = true;
```

Comment:
I am not sure about these properties. The DAG opcode has `SDNPHasChain, SDNPMayLoad, SDNPMayStore`. The intrinsic has `DefaultAttrs` and `IntrInaccessibleMemOrArgMemOnly` and `ReadOnly<ArgIndex<0>>`.

Reply:
I do not really think mayStore is applicable, but for the compatibility with SDag it needs to match. hasSideEffects is probably OK just to reduce possibilities of reordering, but deserves a comment about this, similar to the description of the IntrInaccessibleMemOrArgMemOnly on the intrinsic.

Reply:
should probably just be hasSideEffects. mayLoad/mayStore imply it needs a memory operand and is an ordered memory reference when it doesn't have one

Reply:
I guess it does point to a load address, so no mayStore

Reply:
I have now copied the logic from SDAG that gives it an MMO marked either MOLoad or MOStore depending on the RW flag. I guess that explains why the opcode needs to be both mayLoad and mayStore - unless we split it into separate G_PREFETCH_LOAD/STORE opcodes.

Reply:
> should probably just be hasSideEffects. mayLoad/mayStore imply it needs a memory operand and is an ordered memory reference when it doesn't have one

I could argue this is not a memory operation at all as it shall have no visible effects other than access speed, although practically it has ordering. You certainly do not want a prefetch to be moved past the loads which it was supposed to prefetch. I.e. in my view use of both mayLoad and mayStore is justified. Although we need to make sure it is not considered an aliased store or load from the AA point of view.